---
title: "Infer ASVs with DADA2"
author: "Erica Sawyer"
output: html_document
date: "'r Sys.Data()`"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "/local/workdir/es986/git_repos/SalinityGradient_16S/figures/01_DADA2/")
```


# Goals of this File

1. Use raw fastqc
2. filter and trim bad sequences
3. write out fastqs with high quality sequences
4. eval quality from filter and trim

#Before you start
```{r Set Seed}
# Any number can be chosen
set.seed(958430)
```

# load libraries
``` {r load libraries}

#install.packages("devtools")
library(devtools)

#devtools::install_github("benjjneb/dada2")
library(dada2)


```

# Load Data
``` {r Load Data}
# Set the raw fastq path to the raw sequencing files
# Path to the fastq files

raw_fastq_path <- "/local/workdir/es986/git_repos/SalinityGradient_16S/data/01_DADA2/01_raw_gzipped_fastqs"

# what files in this path
list.files(raw_fastq_path)

# how many files are there
str(list.files(raw_fastq_path))

#create vector or forward reads
forward_reads <- list.files(raw_fastq_path, pattern = "R1_001.fastq.gz", full.names = TRUE)
#intuition check
head(forward_reads)

#create vector of reverse reads
reverse_reads <- list.files(raw_fastq_path, pattern = "R2_001.fastq.gz", full.names = TRUE)
#intuition check
head(reverse_reads)

```

#Quality Plots
```{r raw quality plot}
# randomly select two samples from dataset to evaluate

random_samples <- sample(1:length(reverse_reads), size=2)

# calculate quality and plot of these two samples
plotQualityProfile(forward_reads[random_samples])
plotQualityProfile(reverse_reads[random_samples])


```

# prepare a placeholder for filtered reads
```{r prep filtered sequences}
# vector of our samples, extract sample name from files

samples <- sapply(strsplit(basename(forward_reads),"_"), `[`,1)
#check
head(samples)

# prepare a placeholder for filtered reads
filtered_fastqs_path <- "/local/workdir/es986/git_repos/SalinityGradient_16S/data/01_DADA2/02_filtered_fastqs"

# create 2 variables for forward and reverse filtered

filtered_forward_reads <- file.path(filtered_fastqs_path, paste0(samples, "_R1_filtered.fastq.gz"))
head(filtered_forward_reads)

filtered_reverse_reads <- file.path(filtered_fastqs_path, paste0(samples, "_R2_filtered.fastq.gz"))
head(filtered_reverse_reads)

```



#filter and trim reads

Parameters of filter and trim  **depend on the dataset**
```{r Filter and Trim}
# prepare a placeholder for filtered reads

# maxN = number of N bases. Remove all Ns from data

#maxEE = quality filtering threshold applied to expected errors. Here is there's two expected errors it's ok. If more than two, throw away the sequence. Two values, first for fwd reads, second for rev reads

#trimLeft for the low quality base position

#multithread, compress helps computer

# Assign a vector to filtered reads
# Write out filtered fastq files
filtered_reads <- filterAndTrim(fwd = forward_reads, 
              filt = filtered_forward_reads, 
              rev = reverse_reads, filt.rev = filtered_reverse_reads,
              maxN = 0, maxEE = c(2,2), trimLeft = 3,
              truncQ = 2, rm.phix = TRUE, compress = TRUE)
              #multithread=TRUE for your personal project

#filterAndTrim()

```

# Session Information
```{r Session Info}
# Ensure reproducibility
devtools::session_info()
```