---
title: "Phyloseq PreProcessing"
author: "Erica Sawyer"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Setup
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "/local/workdir/es986/git_repos/SalinityGradient_16S/figures/02_PreProcessing/")
```

# Load Libraries
```{r Load Libraries, echo=FALSE}
#BiocManager::install("phyloseq")
pacman::p_load(devtools, phyloseq, tidyverse, dada2, install=FALSE)
```



#Goals

Here we will process the data into a phyloseq object.

-ASV Table
-Taxonomy Table
-Track Reads (metadata)

Then, we will remove the following:

1. Remove chloroplasts
2. Remove mitochondira
3. Remove samples without 'enough' reads
4. Evaluate accuracy by looking at the mock community
5. Remove negative control taxa if needed


Finally write data file of phyloseq output, which will be stored in data/02_Preprocessing/raw_preprocessed_phyloseq.RData


# Load Data
## ASV Table
```{r Load ASV Table}
# First load ASV Table
load("data/01_DADA2/ASV_counts.RData")

# Inspect asv_tab
head(asv_tab)[,1:5]

# Fix names 
sample_names <- colnames(asv_tab)
samples_fixed <- sapply(strsplit(basename(sample_names), "_"), `[`,1) 
head(samples_fixed)

# re-write the ASV count file to fix names
colnames(asv_tab) <- samples_fixed
str(asv_tab)

```

## Taxonomy Table
```{r Load Taxonomy Table}
tax_df <- read.table("data/01_DADA2/ASV_taxonomy.tsv", sep = "\t", skip=1)
head(tax_df)

# fix column names
colnames(tax_df) <- c("asv_names", "Kingdom", "Phylum", "Class", "Order", "Family", "Genius", "Species", "ASV", "ASVseq")

head(tax_df)
# Taxonomy table matrix

tax_mat <- 
  tax_df %>%
  tibble::column_to_rownames(., var = "asv_names") %>%
  as.matrix()

```

## Track Reads Table
```{r Load Track Reads Table}
load("data/01_DADA2/track_read_counts.RData")

#Take a look at data
head(track_counts_df)
dim(track_counts_df) # 95 rows, 8 col

# load in metadata

metadata_df <- read.csv("data/metadata.csv")

head(metadata_df) 
# view(metadata_df)

dim(metadata_df)
colnames(metadata_df)

head(track_counts_df)

# merge metadata df with track counts df

metadata_track_reads_df <- 
  metadata_df %>%
  left_join(., track_counts_df, by = "names") %>%
  dplyr::select(-X)
  
#check
head(metadata_track_reads_df)

# check row names
row.names(metadata_track_reads_df)

# reassign row names
row.names(metadata_track_reads_df) <- metadata_track_reads_df$names

# check again
row.names(metadata_track_reads_df)

head(metadata_track_reads_df)



```

# Handoff to Phyloseq
```{r Phyloseq handoff}
# double check its all good
dim(asv_tab)
dim(tax_mat) # rows match, that's good

# Intuition check
stopifnot(row.names(asv_tab) == row.names(tax_mat)) #if no output, that means the names are indeed the same; == checks if true or false

# construct phyloseq object

raw_physeq <- phyloseq(otu_table(asv_tab, taxa_are_rows = TRUE),
                       sample_data(metadata_track_reads_df),
                       tax_table(tax_mat))
raw_physeq

# Save this raw phyloseq object
save(raw_physeq, file = "data/02_PreProcessing/raw_physeq.RData")

```


# Clean Up the Data

Remove:
1. Chloroplasts
2. Mitochondria

```{r rm mito and chloro}
# Remind myself of tax table
view(tax_mat) # see where mito and chloro are

# Make raw phyloseq without chloroplasts
noChloro_physeq <-
raw_physeq %>%
  subset_taxa(Order != "Chloroplast" | is.na(Order))

# How many taxa were chloroplasts
num_chloro_ASVs <- ntaxa(raw_physeq) - ntaxa(noChloro_physeq)

# Intuition check
# get from Mar

# Remove mitochondria
noChloroMito_physeq <- 
  noChloro_physeq %>%
  subset_taxa(Family != "Mitochondria" | is.na(Family))
# How many mito asvs
num_mito_ASVs <- ntaxa(noChloro_physeq) - ntaxa(noChloroMito_physeq)
num_mito_ASVs

# How many total asvs removed

ntaxa(raw_physeq) - ntaxa(noChloroMito_physeq)

# proportion of asvs kept
ntaxa(noChloroMito_physeq)/ntaxa(raw_physeq)

```

# Evaluate and remove the control samples

Take a look at the regative controls and then make a decision about whether or not to mreove the ASVs that we found in our controls

1. Negative controls
2. ASVs found within the negative controls and their distribution i the samples
3. Evaluate mock community
```{r neg controls}
# create a vector of samples that were negative controls
control_samples <- c("WaterControl", "022um-Control", "3um-Control", "DNA-Ext-Control")

# make phyloseq object of only the control samples

control_physeq <-
  noChloro_physeq %>%
  subset_samples(., names %in% control_samples) %>%
  # force remove ASVs with 0 counts
  prune_taxa(taxa_sums(.) >0, .)

control_physeq

# vector of control ASVs
 control_ASVs <-
   control_physeq %>%
   tax_table() %>%
   data.frame() %>%
   dplyr::select(ASV) %>%
   as.vector()

# visualize raw abundance values
 control_physeq %>%
   plot_bar(., "ASV", fill= "ASV")

 # check the control asvs in the samples
 #make new phyloseq object with only the control asvs and all samples
 
controlASV_physeq <-
  noChloroMito_physeq %>%
   subset_taxa(., ASV %in% control_ASVs$ASV)
 
# inspect
controlASV_physeq

# look at the abundances
controlASV_physeq %>%
  plot_bar(., "ASV", fill = "ASV", facet_grid = Sample_or_Control~.)

# check what the asvs in the control are
#controlASV_physeq %>%
  #tax_table() %>%
  #data_frame() %>%
 # view()

# only keep asv 18 and 56, everything else needs to be removed

controlASVs_toREMOVE <- dplyr::setdiff(control_ASVs$ASV, c("ASV_18", "ASV_59"))

length(controlASVs_toREMOVE)

## make new phyloseq object
noControlsChlorosMitos_physeq <- 
  noChloroMito_physeq %>%
  # Actually remove the control ASVs (except 18, 56)
  subset_taxa(., !(ASV %in% controlASVs_toREMOVE)) %>%
  # Remove control samples 
  subset_samples(., !(names %in% control_samples))

# Inspect no controls object 
noControlsChlorosMitos_physeq

# After filtering, how many ASVs were removed? 
total_ASVs_rm <- ntaxa(raw_physeq) - ntaxa(noControlsChlorosMitos_physeq)
total_ASVs_rm
```


# Evaluate Accuracy in the Mock Community
-make a mock phyloseq object
-load in the standard fasta sequences of the mock community
-test for the differences between the two
```{r }

# Make mock phyloseq object
mock_physeq <- 
  noControlsChlorosMitos_physeq %>%
  subset_samples(., names =="MockZymoPos") %>%
  prune_taxa(taxa_sums(.) > 0, .)

tax_table(mock_physeq)

# write a mussage sumarizing this
cat("DADA2 inferred", ntaxa(mock_physeq), "ASVs present in the mock community")

### Load in the mock reference seq from zymogen
mock_ref <- getSequences("data/mock_amplicons.fasta")
names(mock_ref)

### Pull out seqs from Sequenced Mock Community
mock_seqs <-
  mock_physeq %>%
  tax_table() %>%
  data.frame() %>%
  dplyr::select(ASV, ASVseq)
#Inspect
head(mock_seqs)

# Which ASVs match the reference
matched_ASVs <-
  mock_seqs %>%
  rowwise() %>%
  # check each ASV if the match (TRUE), if not (FALSE)
  mutate(Match = any(grepl(ASVseq, mock_ref))) %>%
  pull(ASV) #create vector of names that matched

#Evaluate what doesn't match, their abundances, provide error rate





cat(length(matched_ASVs), "ASVs were the exact matches to the expected reference sequences")

# How many sequences were in the mock?
mock_SeqDepth <-
  mock_physeq %>%
  otu_table() %>%
  data.frame() %>%
  colSums()

# Who is in the sequenced mock community: this part is messed up
#mock_physeq %>%
  #tax_table() %>%
 # data.frame() %>%
  #dplyr::select(Species)

# What are the mock counts in the actual samples
noControlsChlorosMitos_physeq %>%
  subset_taxa(., ASV %in% matched_ASVs) %>%
  otu_table() %>%
  t()

```


```{r rm mock asvs}
# Make a new phyloseq object without mock and its asvs
noControlsMock_ChlorosMitos_physeq <-
  noControlsChlorosMitos_physeq %>%
  subset_samples(.,names != "MockZymoPos") %>%
  subset_taxa(., !(ASV %in% matched_ASVs))

num_ASVs_mock <-
  ntaxa(noControlsChlorosMitos_physeq)-ntaxa((noControlsMock_ChlorosMitos_physeq))

#check - you missed this


```

#Evaluate the sequencing depth

```{r Seq Depth}
# The current data object
noControlsMock_ChlorosMitos_physeq

#What is the library size for each sample
seqSums_df <-
  noControlsMock_ChlorosMitos_physeq %>%
  otu_table() %>%
  colSums() %>%
  data.frame() %>%
  rownames_to_column(var = "names") %>%
  left_join(., metadata_track_reads_df, by = "names")
  
dim(seqSums_df)
head(seqSums_df)

# rename column names
colnames(seqSums_df)[2] <- "TotalSeqs"

# show the depth of samples
seqSums_df %>%
  dplyr::select(names, TotalSeqs) %>%
  arrange(TotalSeqs) %>%
  head()


# plot it
seqSums_df %>%
  ggplot(aes(x=reorder(names, TotalSeqs), y=TotalSeqs,
         fill = station)) +
  geom_bar(stat = "identity")

# density plot
seqSums_df %>%
  ggplot(aes(TotalSeqs, fill = station)) +
  geom_density(alpha = 0.5)

```

# Remove samples with few reads
```{r rm samples}

# min reads
min(sample_sums(noControlsMock_ChlorosMitos_physeq))

# remove 20210615-MA-SCB2F
raw_preprocessed_physeq <-
  noControlsMock_ChlorosMitos_physeq %>%
  subset_samples(., names != "20210615-MA-SCB2F")

# check
min(sample_sums(noControlsMock_ChlorosMitos_physeq))


```


#save
```{r save}
save(raw_preprocessed_physeq,
     file = "data/02_PreProcessing/raw_preprocessed_physeq.RData")

```


# Session Information 
```{r session-info}
# Ensure reproducibility 
devtools::session_info()
```
```
