---
title: "Phylogenetic Tree Construction"
author: "Erica Sawyer"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.path = "../figures/03_Phylogenetic_Tree/")
```

# Goals of this File

The goal is to create a phylogenetic tree

1. Load in preprocessed phyloseq object
2. Create ASV fasta file from the phyloseq object
3. Align the 16S sequences from the fasta file with MAFFT
4. Create a tree with FastTree2


## Load Packages & Phyloseq Object

```{r load environment}
#phytools, ggtree, RColorBrewer
pacman::p_load(phyloseq, phytools, ggtree, RColorBrewer, install=FALSE)

#load phyloseq object
load("data/02_PreProcessing/raw_preprocessed_physeq.RData")
raw_preprocessed_physeq

set.seed(238428)
```

# Create fasta file of ASVs and their sequences

This fasta file will be used to create our alignment in MAFFT

```{r create-preprocessed-fasta}
# pull out ASV seqs and ASV names 
asv_seq_df <- 
  raw_preprocessed_physeq@tax_table %>%
  data.frame() %>%
  dplyr::select(ASV, ASVseq)

#View(asv_seq_df)

# Add the > to make fasta header 
asv_seq_df$ASV <- paste0(">",asv_seq_df$ASV)
#View(asv_seq_df)

# Create the fasta object 
asv_seq_fasta <- c(rbind(asv_seq_df$ASV, asv_seq_df$ASVseq))
head(asv_seq_fasta)

# Write to a file 
write(asv_seq_fasta, 
      file = "data/03_Phylogenetic_Tree/preprocessed_ASVs.fasta")
```

# Align the 16S sequences from fasta file with MAFFT

```{r run-mafft, engine='bash', engine.opts='-l'}
# write bash code to run mafft

# first provide the path to mafft
export PATH=/programs/mafft/bin:$PATH 

# change directories to provide fasta file we made above
cd data/03_Phylogenetic_Tree/
pwd

# set seed - using same seed as before for consistency
RANDOM=238428

# Run mafft
# to test in the shell directly from Rmd, command + option + enter (macs)
# For now, use default options
# MAFFT automatically knows to do nucleotide alignment

/programs/mafft/bin/mafft --auto preprocessed_ASVs.fasta > MAFFT_aligned_ASVs.fasta

# Change back to project directory

cd ../../
pwd
```

# FastTree2
```{r run-fasttree, engine='bash', engine.opts='-l'}
# provide export path to fasttree

export PATH=/programs/FastTree-2.1.11:$PATH

# cd into alignment file folder
cd data/03_Phylogenetic_Tree/
pwd

# run fasttree to generate phylogenetic tree
# parameters:
  # -nt = its a nucleotide alignment
  # -gtr = generalized time reversible model (nucleotide substitution model)
  # -fastest = speed up model, reduce memory usage, recommended for datasets with greater than 50k sequences
  # -log = output log file
  # input alignment file
  # specifiy output tree file
  
FastTree -nt -gtr -fastest -log FastTree.log MAFFT_aligned_ASVs.fasta > ASVs_unrooted.tree

cd ../../
# working directory check
pwd
```

# Session Information 
```{r session-info}
# Ensure reproducibility 
devtools::session_info()
```
