---
title: "Between Sample (Beta) Diversity of Microbes Along a Salinity Gradient"
author: "Erica Sawyer"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.path = "../figures/05_CommunityAnalysis/")
```

# Goals

1. Load in phyloseq data with rooted tree
2. Evaluate sequencing depth and remove a sample
3. Normalize the read counts between samples
4. Calculate community dissimilarities. Numbers between 0 and 1. If 0: completely similar; if 1, completely dissimilar
  a. Sorenson- Shared species by a binary value (abundance unweighted)
  b. Bray-Curtis- Shared Abundance species (abundance weighted)
  c. (abundance) Weighted UNIFRAC- Consider abundant species and where they fall on the tree
5. Visualize the community data with two unconstrained Ordinations
  a. PCoA- Linear method = Eigenvalue = Calculate how much variation is explained by each axis. Chooe to view axis 1, 2, 3, etc and plot them together
  b. NMDS- Non-linear methods: Smush multiple dimensions into 2 or 3 axes. Need to report stress value (Ideally <0.15)
6. Run statistics with permanova and betadi.spR

  
# Setup

## Load Libraries
```{r}
#install.packages("vegan")
pacman::p_load(tidyverse, devtools, patchwork, vegan, phyloseq, 
               install = FALSE)
# Define Colors
station_colors <- c(
  "Shipping Channel" = "dodgerblue4",
  "Aransas Bay" = "dodgerblue2",
  "Copano West" = "#D9CC3C",
  "Copano East" = "#A0E0BA",
  "Mesquite Bay" = "#00ADA7")

```
## Load Data
```{r load data}
# load in rooted phylogenetic tree
load("data/03_Phylogenetic_Tree/phytree_preprocessed_physeq.RData")
midroot_physeq_rm456
unrooted_physeq_rm476
```


## Explore Read Counts - Raw Reads
```{r calculate sequence depth}
# Calculate the total number of reads per sample
raw_TotalSeqs_df <- 
  midroot_physeq_rm456 %>%
  # calcualte the sample read sums
  sample_sums() %>%
  data.frame() 

# name the columns
colnames(raw_TotalSeqs_df)[1] <- "TotalSeqs"

head(raw_TotalSeqs_df)

# Make histogram of raw reads
raw_TotalSeqs_df %>%
  ggplot(aes(x =TotalSeqs)) +
  geom_histogram(bins=50) +
  scale_x_continuous(limits = c(0, 10000)) +
  labs(title= "Raw Sequencing Depth Distribution") +
  theme_classic()

```

## Remove low seq sample
```{r remove low seq sample}
raw_rooted_physeq <-
  midroot_physeq_rm456 %>%
  # remove low seq sample that was alpha diversity outlier
  subset_samples(names != "20210615-MA-ABB2F") %>%
  prune_taxa(taxa_sums(.) > 0, .)


# inspect
raw_rooted_physeq

# what is sequence read minimun
raw_rooted_physeq %>%
  sample_sums() %>%
  min()
```

## Normalize Read Counts
```{r scale reads}
#scale_reads function
# Function to scale reads: http://deneflab.github.io/MicrobeMiseq/ 
# Scales reads by 
# 1) taking proportions
# 2) multiplying by a given library size of n
# 3) rounding 
# Default for n is the minimum sample size in your library
# Default for round is floor

matround <- function(x){trunc(x+0.5)}

scale_reads <- function(physeq, n = min(sample_sums(physeq)), round = "round") {
  
  # transform counts to n
  physeq.scale <- transform_sample_counts(physeq, function(x) {(n * x/sum(x))})
  
  # Pick the rounding functions
  if (round == "floor"){
    otu_table(physeq.scale) <- floor(otu_table(physeq.scale))
  } else if (round == "round"){
    otu_table(physeq.scale) <- round(otu_table(physeq.scale))
  } else if (round == "matround"){
    otu_table(physeq.scale) <- matround(otu_table(physeq.scale))
  }
  
  # Prune taxa and return new phyloseq object
  physeq.scale <- prune_taxa(taxa_sums(physeq.scale) > 0, physeq.scale)
  return(physeq.scale)
}

```


## Scale the reads and check the distribution of the read seq depth

This is where you would decide to use rarefaction to normalize data
```{r scale physeq}
min(sample_sums(raw_rooted_physeq))

scaled_rooted_physeq <-
  raw_rooted_physeq %>%
  scale_reads(round = "matround")

# Calculate the read depth
scaled_TotalSeqs_df <-
  scaled_rooted_physeq %>%
  sample_sums() %>%
  data.frame()

colnames(scaled_TotalSeqs_df)[1] <-"TotalSeqs"

# Inspect
head(scaled_TotalSeqs_df)

# Check the range of the data
minseqs <- min(scaled_TotalSeqs_df)
maxseqs <- max(scaled_TotalSeqs_df)

# range
maxseqs - minseqs

# Plot histogram
scaled_TotalSeqs_df %>%
  ggplot(aes(x =TotalSeqs)) +
  geom_histogram(bins=50) +
  scale_x_continuous(limits = c(0, 10000)) +
  labs(title= "Scaled Sequencing Depth Distribution at 2194") +
  theme_classic()
```

# Calculate & Visualize Community Dissimilarity

Exploratory Analysis from the Paliy and Shankar (2016) paper, which is unconstranied ordiantion methods like PCoA
## Sorenson PCoA

```{r sorenson PCoA}
# calculate sorneson dissimilarity
scaled_soren_pcoa <-
  ordinate(
  physeq = scaled_rooted_physeq,
  method = "PCoA",
  distance ="bray", binary = TRUE)

str(scaled_soren_pcoa)

# plot ordination
plot_ordination(
  physeq = scaled_rooted_physeq,
  ordination = scaled_soren_pcoa,
  color = "station",
  title = "Sorenson PCoA") + 
  scale_color_manual(values = station_colors) +
  theme_bw()

# Permanova of Sorenson
# make a new object that has sorenson dissimilarity matrix
sorenson_distance <-
  phyloseq::distance(scaled_rooted_physeq, method = "bray", binary = TRUE)

str(sorenson_distance)

# metadata
metadata <- 
  scaled_rooted_physeq %>%
  sample_data() %>%
  data.frame()

# run permanova
# Testing if the centroides of the data are similar or different
adonis2(sorenson_distance ~ station, data = metadata) 

```

# Taxanomic Composition
```{r taxanomic composition}

# set phylum colors
phylum_colors <- c(
  Acidobacteriota = "navy", 
  Actinobacteriota = "darkslategray2", 
  Armatimonadota = "deeppink1",
  Alphaproteobacteria = "plum2", 
  Bacteroidota = "gold", 
  Betaproteobacteria = "plum1", 
  Bdellovibrionota = "red1",
  Chloroflexi="black", 
  Crenarchaeota = "firebrick",
  Cyanobacteria = "limegreen",
  Deltaproteobacteria = "grey", 
  Desulfobacterota="magenta",
  Firmicutes = "#3E9B96",
  Gammaproteobacteria = "greenyellow",
  "Marinimicrobia (SAR406 clade)" = "yellow",
  Myxococcota = "#B5D6AA",
  Nitrospirota = "palevioletred1",
  Proteobacteria = "royalblue",
  Planctomycetota = "darkorange", 
  "SAR324 clade(Marine group B)" = "olivedrab",
  #Proteobacteria_unclassified = "greenyellow",
  Thermoplasmatota = "green",
  Verrucomicrobiota = "darkorchid1")
 # Other = "grey")

# Calculate relative abundance
# Note: The read depth must be normalized in some way

phylum_df <-
  scaled_rooted_physeq %>%
  # agglomerate at the phylum level
  tax_glom(taxrank = "Phylum") %>%
  # transform the counts to relative abundance
  transform_sample_counts(function (x) {x/sum(x)}) %>%
  # melt to a long format
  psmelt() %>%
  # filter out phyla less than 1%
  dplyr::filter(Abundance > 0.01) %>%
  mutate(date = fct_relevel(date, c("6/2/21", "6/15/21", "10/5/21")),
         station = fct_relevel(station, c("Copano West", "Copano East",
                                          "Mesquite Bay", "Aransas Bay",
                                          "Shipping Channel")))
  

# Plot Phylum Abundances - Stacked Bar Plot with All Phyla
phylum_df %>%
  # you need one sample per x value or it will take the sum of multiple samples
  dplyr::filter(depth == 0.0) %>%
  dplyr::filter(fraction == "Whole") %>%
  ggplot(aes(x = station, y = Abundance, fill = Phylum)) +
  facet_grid(.~date) + # fix this line
  geom_bar(stat = "identity", color = "black") +
  labs(Title = "Surface Phylum Composistion") +
  scale_fill_manual(values = phylum_colors) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1))

# Make each phyla its own row
phylum_df %>%
  # you need one sample per x value or it will take the sum of multiple samples
  dplyr::filter(depth == 0.0) %>%
  dplyr::filter(fraction == "Whole") %>%
  ggplot(aes(x = station, y = Abundance, fill = Phylum)) +
  facet_grid(Phylum~date, scales = "free") + # fix this line
  geom_bar(stat = "identity", color = "black") +
  labs(Title = "Surface Phylum Composistion") +
  scale_fill_manual(values = phylum_colors) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1))

# Narrow in on a specific group
# Actinobacteroidota: y=abundance, x=station, dot plot plus box plot
phylum_df %>%
  dplyr::filter(Phylum == "Actinobacteriota") %>%
  # Built this plot
  ggplot(aes(x = station, y = Abundance, fill = station)) +
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here
  geom_jitter() +
  theme_bw() +
  labs(title = "Actinobacteriota Phylum Abundance") +
  scale_color_manual(values = station_colors) +
  scale_fill_manual(values = station_colors)

# For stats: Do Kruskall-Wallis followed by a Tukey Posthoc Test; these are nonparametric stat tests
```


## Family Composition
```{r}
family_df <-
  scaled_rooted_physeq %>%
  # agglomerate at the phylum level
  tax_glom(taxrank = "Family") %>%
  # transform the counts to relative abundance
  transform_sample_counts(function (x) {x/sum(x)}) %>%
  # melt to a long format
  psmelt() %>%
  # filter out phyla less than 1%
  dplyr::filter(Abundance > 0.01) %>%
  mutate(date = fct_relevel(date, c("6/2/21", "6/15/21", "10/5/21")),
         station = fct_relevel(station, c("Copano West", "Copano East",
                                          "Mesquite Bay", "Aransas Bay",
                                          "Shipping Channel")))
  

# Actinobacteroidota: y=abundance, x=station, dot plot plus box plot
family_df %>%
  dplyr::filter(Phylum == "Actinobacteriota") %>%
  # Built this plot
  ggplot(aes(x = station, y = Abundance, fill = station)) +
  facet_wrap(.~Family, scales ="free_y", nrow = 1) +
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here
  geom_jitter() +
  theme_bw() +
  labs(title = "Actinobacteriota Family Abundance") +
  scale_color_manual(values = station_colors) +
  scale_fill_manual(values = station_colors)


```

## Bray-Curtis PCoA

## Weighted UniFrac PCoA

## Bray-Curtis NMDS

# Test for Statistical Significance with PERMANOVA & betadi.spR

# Session Information 
```{r session-info}
# Ensure reproducibility 
devtools::session_info()
```
